### PR TITLE
restructure validation PP 

### LINKED ISSUE #1652: [TASK] Validation Postprocessor should be restructured to use the new postprocessor API in postprocessorreadyinterface
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
Generalize the Validation PostProcessor base class to make it easier to create new validation postprocessor.  

**Describe the solution you'd like**
Restructure the validation postprocessor base class, let it inherit from the PostProcessorReadyInterface, utilize the DataSet directly for the validation computation. 

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1652 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by wangcj05** [comment]
@Jimmy-INL @yoshiurr-INL This the PR for you. The remaining things for you to do is to merge the two "run" methods. Please let me know if you have any question.

File: framework/Metrics/metrics/RepresentativityFactors.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import copy
+#import scipy.spatial.distance as spatialDistance
+#External Modules End--------------------------------------------------------------------------------
+
+#Internal Modules------------------------------------------------------------------------------------
+from .Metric import Metric
+from utils import InputData, InputTypes
+#Internal Modules End--------------------------------------------------------------------------------
+
+class RepresentativityFactors(Metric):
```
Comment (wangcj05) [inline framework/Metrics/metrics/RepresentativityFactors.py:34]: I suggest to move this to another PR

File: framework/Models/PostProcessors/validationAlgorithms/Representativity.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+from utils import InputData, InputTypes
+#import Files
+#import Distributions
+#import MetricDistributor
+from utils import utils
+from .ValidationBase import ValidationBase
+from utils.mathUtils import partialDerivative, derivatives
+#Internal Modules End--------------------------------------------------------------------------------
+
+class Representativity(ValidationBase):
```
Comment (wangcj05) [inline framework/Models/PostProcessors/validationAlgorithms/Representativity.py:39]: I suggest to move this file to another PR.

File: tests/framework/PostProcessors/Validation/test_validation_gate_representativity.xml [hunk 3]
```diff
@@ -0,0 +1,117 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="all">
+  <RunInfo>
```
Comment (wangcj05) [inline tests/framework/PostProcessors/Validation/test_validation_gate_representativity.xml:3]: Same here, I suggest to move this file to another PR

File: framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py [hunk 4]
```diff
@@ -135,22 +71,121 @@ def _handleInput(self, paramInput):
...
+    # for feat, targ in zip(self.features, self.targets):
+    #   featData = self._getDataFromDatasets(dataSets, feat, names)
+    #   targData = self._getDataFromDatasets(dataSets, targ, names)
+    #   for metric in self.metrics:
+    #     name = "{}_{}_{}".format(feat.split("|")[-1], targ.split("|")[-1], metric.estimator.name)
+    #     outs[name] = metric.evaluate((featData, targData), multiOutput='raw_values')
+    # return outs
+
+    names=[]
+    if isinstance(inputIn['Data'][0][-1], xr.Dataset):
```
Comment (wangcj05) [inline framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py:98]: This check is not needed

File: framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py [hunk 5]
```diff
@@ -135,22 +71,121 @@ def _handleInput(self, paramInput):
...
+
+  def run(self, inputIn):
+    """
+      This method executes the postprocessor action. In this case it loads the
+      results to specified dataObject
+      @ In, inputIn, list, dictionary of data to process
+      @ Out, outputDict, dict, dictionary containing the post-processed results
+    """
+    # inpVars, outVars, dataSet = inputIn['Data'][0]
+    dataSets = [data for _, _, data in inputIn['Data']]
```
Comment (wangcj05) [inline framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py:83]: I suggest to store the dataSets into a dictionary, i.e.

dataDict = {data.attrs['name']: data for _, _, data in inputIn['Data']}

In this case you can get ride of _evaluate method, and much easier to retrieve the feat/targ data from the dataDict.

**Review by wangcj05 — COMMENTED** [review]
Some comments for your consideration

**Comment by wangcj05** [comment]
@Jimmy-INL Any update on this?

**Comment by wangcj05** [comment]
@Jimmy-INL Any update on this?

**Review by Jimmy-INL — APPROVED** [review]

File: framework/DataObjects/DataSet.py [hunk 6]
```diff
@@ -2065,18 +2066,19 @@ def _toCSVXML(self,fileName,**kwargs):
...
-          for var in child.text.split(','):
-            if var.strip() in keep:
-              vs.append(var)
-          if len(vs) == 0:
-            toRemove.append(child)
-          else:
-            child.text = ','.join(vs)
-      for r in toRemove:
-        genNode.remove(r)
+      if genNode is not None:
```
Comment (mandd) [inline framework/DataObjects/DataSet.py:2069]: what is the rationale behind this line being added here?

File: framework/Models/PostProcessors/Factory.py [hunk 7]
```diff
@@ -36,6 +36,7 @@
 from .ParetoFrontierPostProcessor import ParetoFrontier
 from .EconomicRatio import EconomicRatio
 from .Validation import Validation
+from .validationAlgorithms.Probabilistic import Probabilistic
```
Comment (mandd) [inline framework/Models/PostProcessors/Factory.py:39]: should this be from .ValidationAlgorithms.Probabilistic import Probabilistic

File: framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py [hunk 8]
```diff
@@ -23,14 +23,15 @@
 
 #External Modules------------------------------------------------------------------------------------
 import numpy as np
+import xarray as xr
 #External Modules End--------------------------------------------------------------------------------
 
 #Internal Modules------------------------------------------------------------------------------------
 from utils import utils
-from .ValidationBase import ValidationBase
+from ..Validation import Validation
```
Comment (mandd) [inline framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py:31]: should this be from ".Validation import Validation"?

**Review by mandd — CHANGES_REQUESTED** [review]

File: framework/Models/PostProcessors/Factory.py (hunk 7 above)
Comment (joshua-cogliati-inl) [reply framework/Models/PostProcessors/Factory.py:39]: Hm, we would have to rename validationAlgorithms to ValidationAlgorithms (currently in RAVEN it is validationAlgorithms).

File: tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml [hunk 9]
```diff
@@ -25,17 +25,14 @@
     <ExternalModel ModuleToLoad="../../AnalyticModels/atten_and_poly.py" name="poly" subType="">
       <variables>x1,x2,ans,ans2</variables>
     </ExternalModel>
-    <PostProcessor name="pp1" subType="Validation">
+    <PostProcessor name="pp1" subType="Probabilistic">
```
[the remark this answers is not in the document]
Comment (joshua-cogliati-inl) [reply tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml:28]: Hm, does the user manual need to be updated?

File: framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py (hunk 8 above)
Comment (Jimmy-INL) [reply framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py:31]: No, I think it needs two dots as it is in a parent folder.

File: framework/Models/PostProcessors/Factory.py (hunk 7 above)
Comment (Jimmy-INL) [reply framework/Models/PostProcessors/Factory.py:39]: resolved

File: framework/DataObjects/DataSet.py (hunk 6 above)
Comment (Jimmy-INL) [reply framework/DataObjects/DataSet.py:2069]: @wangcj05 ?

File: framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py (hunk 8 above)
Comment (Jimmy-INL) [reply framework/Models/PostProcessors/validationAlgorithms/Probabilistic.py]: I do not think so. Validation is two folders up. It is located in PostProsessors not in validationAlgorithms

File: framework/Models/PostProcessors/Factory.py (hunk 7 above)
Comment (Jimmy-INL) [reply framework/Models/PostProcessors/Factory.py]: I have renamed it to ValidationAlgorithms

File: framework/DataObjects/DataSet.py (hunk 6 above)
Comment (wangcj05) [reply framework/DataObjects/DataSet.py:2069]: This is just to make sure the point wise meta is presented in the dataset.

File: tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jimmy-INL) [reply tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml:28]: I have corrected this in the manual

File: framework/Models/PostProcessors/Validation.py [hunk 10]
```diff
@@ -88,8 +86,22 @@ def __init__(self):
...
+
+    self.dataType = ['static', 'dynamic'] # the type of data can be passed in (static aka PointSet, dynamic aka HistorySet) (if both are present the validation algorithm can work for both data types)
+    self.acceptableMetrics = [] # if not populated all types of metrics are accepted, otherwise list the metrics (see Probablistic.py for an example)
+    self.features = None        # list of feature variables
+    self.targets = None         # list of target variables
+    self.pivotValues = None     # pivot values (present if dynamic == True)
+
+
     self.addAssemblerObject('Metric', InputData.Quantity.one_to_infinity)
     self.addAssemblerObject('PreProcessor', InputData.Quantity.zero_to_infinity)
```
Comment (Jimmy-INL) [inline framework/Models/PostProcessors/Validation.py:98]: @wangcj05 can we make this zero_to_infinity? Some validation methods might not have a metric.

File: framework/Models/PostProcessors/Validation.py [hunk 11]
```diff
@@ -192,71 +156,61 @@ def _handleInput(self, paramInput):
...
-  def collectOutput(self, finishedJob, output):
+  def initialize(self, runInfo, inputs, initDict):
     """
-      Function to place all of the computed data into the output object
-      @ In, finishedJob, JobHandler External or Internal instance, A JobHandler
-        object that is in charge of running this post-processor
-      @ In, output, dataObjects, A reference to an object where we want
-        to place our computed results
+      Method to initialize the DataMining pp.
+      @ In, runInfo, dict, dictionary of run info (e.g. working dir, etc)
```
Comment (Jimmy-INL) [inline framework/Models/PostProcessors/Validation.py:169]: @wangcj05 Shouldn't the docstring be updated? validation PP instead of DataMining PP?

File: framework/Models/PostProcessors/Validation.py (hunk 10 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/Validation.py:98]: I think so. Go ahead and change it.

File: framework/Models/PostProcessors/Validation.py (hunk 11 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/Validation.py:169]: Yes, go ahead and update it.

**Comment by wangcj05** [comment]
@joshua-cogliati-inl @mandd We have resolved your comments, can you take another look at this PR?

File: framework/Models/PostProcessors/Validations/Probabilistic.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+      @ In, kwargs, dict, keyword arguments
+      @ Out, outputDict, dict, dictionary containing the results {"feat"_"target"_"metric_name":value}
+    """
+    names = kwargs.get('dataobjectNames')
+    outputDict = {}
+    for feat, targ in zip(self.features, self.targets):
+      featData = self._getDataFromDataDict(datasets, feat, names)
+      targData = self._getDataFromDataDict(datasets, targ, names)
+      for metric in self.metrics:
+        name = "{}_{}_{}".format(feat.split("|")[-1], targ.split("|")[-1], metric.estimator.name)
```
Comment (joshua-cogliati-inl) [inline framework/Models/PostProcessors/Validations/Probabilistic.py:113]: Hm, could this feat.split("|")[-1] ever result in overlapping names? (As in if we have bar|baz and bat|baz they will end up overwriting in the dictionary)

Comment (joshua-cogliati-inl) [reply framework/Models/PostProcessors/Validations/Probabilistic.py:113]: Tho' this did exist in the previous version, so without renaming the name, it may not be fixable.

**Review by joshua-cogliati-inl — APPROVED** [review]
No new objections.

**Comment by joshua-cogliati-inl** [comment]
Possibly a conversion script needs to be written since this will change existing inputs.

**Review by joshua-cogliati-inl — APPROVED** [review]
Approve.

File: tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml [hunk 13]
```diff
@@ -25,17 +25,14 @@
...
       <Features>outputDataMC1|ans</Features>
       <Targets>outputDataMC2|ans2</Targets>
       <!--
         <Metric class="Metrics" type="RACDistance">fz_expected</Metric>
       -->
       <Metric class="Metrics" type="Metric">cdf_diff</Metric>
       <Metric class="Metrics" type="Metric">pdf_area</Metric>
-      <Probabilistic name="myProbMetric">
```
[the remark this answers is not in the document]
Comment (joshua-cogliati-inl) [reply tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml:36]: @Jimmy-INL @wangcj05  Is this still a problem?

**Comment by joshua-cogliati-inl** [comment]
> Possibly a conversion script needs to be written since this will change existing inputs.

Thanks for the conversion script.

File: tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (wangcj05) [reply tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml:36]: I do not think this is a problem for validation post processor. Additional nodes can be added directly in each specific postprocessor.

[the remark this answers is not in the document]
Comment (Jimmy-INL) [reply tests/framework/PostProcessors/Validation/test_validation_gate_probabilistic.xml:36]: I am OK with it.

**Review by mandd — APPROVED** [review]

**Comment by mandd** [comment]
Review board and tests are green. Merging PR
